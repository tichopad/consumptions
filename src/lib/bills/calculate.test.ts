import { type EnergyType, type ID } from '$lib/models/common';
import type { Occupant } from '$lib/models/occupant';
import { assertSuccess, sumPreciseBy } from '$lib/utils';
import { assert, describe, expect, test } from 'vitest';
import { CalculationInputError, calculateBills } from './calculate';
import type { MeasuringDeviceConsumption, OccupantCalculationEntry } from './calculation-types';

describe('calculateBills', () => {
	// -- Context --

	const buildingId = 'building_1' as ID;
	const billingPeriodId = 'billing_period_1' as ID;

	const renterA = makeRenter('renterA', buildingId, { squareMeters: 10 });
	const renterB = makeRenter('renterB', buildingId, { squareMeters: 20 });

	const rentersElectricityConsumption = makeMeasurement('renterWithDevice', 'electricity', 100);
	const renterWithDevice = makeRenter('renterWithDevice', buildingId, {
		squareMeters: 30,
		measuringDevices: [rentersElectricityConsumption]
	});

	const ownerAWaterConsumption = makeMeasurement('ownerA', 'water', 2);
	const ownerAElectricityConsumption = makeMeasurement('ownerA', 'electricity', 50.55);
	const ownerAHeatingConsumption = makeMeasurement('ownerA', 'heating', 10);
	const ownerA = makeOwner('ownerA', buildingId, {
		squareMeters: 20,
		measuringDevices: [
			ownerAElectricityConsumption,
			ownerAHeatingConsumption,
			ownerAWaterConsumption
		]
	});

	const ownerBWaterConsumption = makeMeasurement('ownerB', 'water', 3);
	const ownerBElectricityConsumption = makeMeasurement('ownerB', 'electricity', 75.12);
	const ownerBHeatingConsumption = makeMeasurement('ownerB', 'heating', 5);
	const ownerB = makeOwner('ownerB', buildingId, {
		heatingFixedCostShare: 10,
		measuringDevices: [
			ownerBElectricityConsumption,
			ownerBHeatingConsumption,
			ownerBWaterConsumption
		]
	});

	// -- Tests --

	test('calculates bills for renters only', () => {
		const result = calculateBills({
			buildingId,
			billingPeriodId,
			energyType: 'electricity',
			totalConsumption: 1200,
			totalCost: 7200,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA, renterB]
		});

		assert(result.success === true, 'Calculation failed');
		const { billsToInsert, consumptionRecordsToInsert } = result.value;

		expect(result.value).toMatchSnapshot();

		// Association to occupant and building is mutually exclusive
		for (const bill of billsToInsert) {
			if (bill.occupantId) expect(bill.buildingId).toBeFalsy();
			else if (bill.buildingId) expect(bill.occupantId).toBeFalsy();
		}

		// There shouldn't be any consumption records
		expect(consumptionRecordsToInsert).toHaveLength(0);
	});

	test('calculates bills for renters with measuring devices', () => {
		const result = calculateBills({
			buildingId,
			billingPeriodId,
			energyType: 'electricity',
			totalConsumption: 1200,
			totalCost: 7200,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA, renterB, renterWithDevice]
		});

		assert(result.success === true, 'Calculation failed');
		expect(result.value).toMatchSnapshot();
	});

	test('calculates bills for renters, renters with measuring devices and owners', () => {
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1500,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA, renterB, renterWithDevice, ownerA, ownerB]
		});

		assert(result.success === true, 'Calculation failed');
		expect(result.value).toMatchSnapshot();
	});

	test('calculated bills equal the input total cost', () => {
		const electricityResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1500,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA, renterWithDevice, ownerA, ownerB]
		});
		const waterResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'water',
			totalConsumption: 10,
			totalCost: 1000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA, renterWithDevice, ownerA, ownerB]
		});

		assert(electricityResult.success);
		assert(waterResult.success);

		const occupantsElectricityBills = electricityResult.value.billsToInsert.filter(
			(b) => !b.buildingId
		);
		const occupantsElectricityBillsSum = sumPreciseBy(
			occupantsElectricityBills,
			(b) => b.totalCost
		);
		expect(occupantsElectricityBillsSum.toNumber()).toBeCloseTo(9000, 0.1);

		const occupantsWaterBills = waterResult.value.billsToInsert.filter((b) => !b.buildingId);
		const occupantsWaterBillsSum = sumPreciseBy(occupantsWaterBills, (b) => b.totalCost);
		expect(occupantsWaterBillsSum.toNumber()).toBeCloseTo(1000, 0.1);
	});

	test('returns zero cost per square meter when there are no renters', () => {
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 125.67,
			totalCost: 754.02,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [ownerA, ownerB]
		});

		assert(result.success);

		const buildingBill = result.value.billsToInsert.find((b) => b.buildingId === buildingId);

		expect(buildingBill?.costPerSquareMeter).toBe(0);
	});

	test('fails given no occupants', () => {
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1500,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: []
		});
		assert(result.success === false);
		expect(result.error).toBeInstanceOf(CalculationInputError);
	});

	test('fails given invalid total consumption', () => {
		const zeroResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: -420,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(zeroResult.success === false);
		expect(zeroResult.error).toBeInstanceOf(CalculationInputError);

		const negativeResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: -1000,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(negativeResult.success === false);
		expect(negativeResult.error).toBeInstanceOf(CalculationInputError);
	});

	test('fails given invalid total cost', () => {
		const zeroResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1000,
			totalCost: 0,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(zeroResult.success === false);
		expect(zeroResult.error).toBeInstanceOf(CalculationInputError);

		const negativeResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1000,
			totalCost: -9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(negativeResult.success === false);
		expect(negativeResult.error).toBeInstanceOf(CalculationInputError);
	});

	test('fails given invalid fixed cost', () => {
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'heating',
			fixedCost: -1000,
			totalConsumption: 95,
			totalCost: 42500,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(result.success === false);
		expect(result.error).toBeInstanceOf(CalculationInputError);
	});

	test('fails given fixed cost and other energy type than "heating"', () => {
		const waterResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'water',
			fixedCost: 5000,
			totalConsumption: 95,
			totalCost: 42500,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(waterResult.success === false);
		expect(waterResult.error).toBeInstanceOf(CalculationInputError);

		const electricityResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			fixedCost: 5000,
			totalConsumption: 95,
			totalCost: 42500,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assert(electricityResult.success === false);
		expect(electricityResult.error).toBeInstanceOf(CalculationInputError);
	});

	test('handles zero total consumption', () => {
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'heating',
			totalConsumption: 0,
			totalCost: 9000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renterA]
		});
		assertSuccess(result);
		expect(result.value).toMatchSnapshot();
	});

	test.concurrent('correctly splits fixed cost', () => {
		const bId = 'building_1' as ID;
		const occupant1 = makeOwner('occupant1', bId, {
			heatingFixedCostShare: 10,
			measuringDevices: [makeMeasurement('occupant1', 'heating', 10)]
		});
		const occupant2 = makeOwner('occupant2', bId, {
			measuringDevices: [makeMeasurement('occupant2', 'heating', 20)]
		});
		const occupant3 = makeOwner('occupant3', bId, {
			chargedUnmeasuredHeating: true,
			squareMeters: 10,
			measuringDevices: []
		});
		const result = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'heating',
			totalConsumption: 50,
			totalCost: 40000,
			fixedCost: 20000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [occupant1, occupant2, occupant3]
		});
		assertSuccess(result);

		const occupant1Bill = result.value.billsToInsert.find((b) => b.occupantId === occupant1.id);
		expect(occupant1Bill?.costPerUnit).toBe(400);
		expect(occupant1Bill?.totalCost).toBe(24000);
		expect(occupant1Bill?.fixedCost).toBe(20000);
		expect(occupant1Bill?.measuredCost).toBe(4000);

		const occupant2Bill = result.value.billsToInsert.find((b) => b.occupantId === occupant2.id);
		expect(occupant2Bill?.costPerUnit).toBe(400);
		expect(occupant2Bill?.totalCost).toBe(8000);
		expect(occupant2Bill?.fixedCost).toBeUndefined();
		expect(occupant2Bill?.measuredCost).toBe(8000);

		const occupant3Bill = result.value.billsToInsert.find((b) => b.occupantId === occupant3.id);
		expect(occupant3Bill?.totalCost).toBe(8000);
		expect(occupant3Bill?.costPerSquareMeter).toBe(800);
		expect(occupant3Bill?.fixedCost).toBeUndefined();
		expect(occupant3Bill?.measuredCost).toBeUndefined();
	});

	test.concurrent('handles real world calculation scenario', ({ expect }) => {
		// Setup
		const owner1ElectricityConsumption = makeMeasurement('owner1', 'electricity', 150);
		const owner1WaterConsumption = makeMeasurement('owner1', 'water', 3);
		const owner1HeatingConsumption = makeMeasurement('owner1', 'heating', 5);
		const owner1 = makeOwner('owner1', buildingId, {
			heatingFixedCostShare: null,
			measuringDevices: [
				owner1ElectricityConsumption,
				owner1WaterConsumption,
				owner1HeatingConsumption
			]
		});
		const owner2ElectricityConsumption = makeMeasurement('owner2', 'electricity', 200);
		const owner2WaterConsumption = makeMeasurement('owner2', 'water', 6.5);
		const owner2HeatingConsumption = makeMeasurement('owner2', 'heating', 10);
		const owner2 = makeOwner('owner2', buildingId, {
			heatingFixedCostShare: 10,
			measuringDevices: [
				owner2ElectricityConsumption,
				owner2WaterConsumption,
				owner2HeatingConsumption
			]
		});
		const oliera = makeOwner('oliera', buildingId, {
			heatingFixedCostShare: 771,
			measuringDevices: []
		});
		const renter1 = makeRenter('renter1', buildingId, { squareMeters: 20 });
		const renter2 = makeRenter('renter2', buildingId, { squareMeters: 10 });
		const renter3 = makeRenter('renter3', buildingId, { squareMeters: 50 });

		// Electricity
		const electricityResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'electricity',
			totalConsumption: 1550,
			totalCost: 9300,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renter1, renter2, renter3, owner1, owner2]
		});
		assert(electricityResult.success === true);
		expect(electricityResult.value).toMatchSnapshot();

		// Water
		const waterResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'water',
			totalConsumption: 11.6,
			totalCost: 1023.352,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renter1, renter2, renter3, owner1, owner2]
		});
		assert(waterResult.success === true);
		expect(waterResult.value).toMatchSnapshot();

		// Heating - this is where things get interesting since there's a fixed cost share
		const heatingResult = calculateBills({
			billingPeriodId,
			buildingId,
			energyType: 'heating',
			totalConsumption: 95,
			totalCost: 42388.05,
			fixedCost: 5000,
			dateRange: {
				start: new Date('2024-01-01T00:00:00Z'),
				end: new Date('2024-01-31T00:00:00Z')
			},
			occupants: [renter1, renter2, renter3, owner1, owner2, oliera]
		});
		assert(heatingResult.success === true);
		expect(heatingResult.value).toMatchSnapshot();
	});
});

// -- Helpers --

function makeMeasurement(
	label: string,
	energyType: EnergyType,
	consumption: number
): MeasuringDeviceConsumption {
	return {
		id: ('measurement_record_' + label) as ID,
		energyType,
		consumption
	};
}

function makeRenter(
	label: string,
	buildingId: ID,
	values: Partial<OccupantCalculationEntry> = {}
): OccupantCalculationEntry {
	return {
		...makeOccupant(buildingId),
		measuringDevices: [],
		chargedUnmeasuredElectricity: true,
		chargedUnmeasuredHeating: true,
		chargedUnmeasuredWater: true,
		...values,
		id: ('occupant_' + label) as ID
	};
}

function makeOwner(
	label: string,
	buildingId: ID,
	values: Partial<OccupantCalculationEntry> = {}
): OccupantCalculationEntry {
	return {
		...makeOccupant(buildingId),
		measuringDevices: [],
		chargedUnmeasuredElectricity: false,
		chargedUnmeasuredHeating: false,
		chargedUnmeasuredWater: false,
		...values,
		id: ('occupant_' + label) as ID
	};
}

function makeOccupant(buildingId: ID): Occupant {
	return {
		id: `occupant_${buildingId}` as ID,
		buildingId,
		chargedUnmeasuredElectricity: false,
		chargedUnmeasuredHeating: false,
		chargedUnmeasuredWater: false,
		name: 'Occupant',
		squareMeters: 69,
		created: new Date('2023-01-01T00:00:00Z'),
		deleted: null,
		archived: null,
		heatingFixedCostShare: null,
		isDeleted: false,
		isArchived: false,
		updated: new Date('2024-01-01T00:00:00Z')
	};
}
