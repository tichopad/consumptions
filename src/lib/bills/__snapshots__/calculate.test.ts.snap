// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calculateBills > calculates bills for renters only 1`] = `
{
  "billsToInsert": [
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 240,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterA",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 2400,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 240,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterB",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 4800,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 240,
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 1200,
      "totalCost": 7200,
    },
  ],
  "consumptionRecordsToInsert": [],
}
`;

exports[`calculateBills > calculates bills for renters with measuring devices 1`] = `
{
  "billsToInsert": [
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 600,
      "occupantId": "occupant_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 100,
      "totalCost": 600,
    },
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 110,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterA",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 1100,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 110,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterB",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 2200,
    },
    {
      "billedArea": 30,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 110,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 3300,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 110,
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 1200,
      "totalCost": 7200,
    },
  ],
  "consumptionRecordsToInsert": [
    {
      "consumption": 100,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`calculateBills > calculates bills for renters, renters with measuring devices and owners 1`] = `
{
  "billsToInsert": [
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 600,
      "occupantId": "occupant_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 100,
      "totalCost": 600,
    },
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 303.3,
      "occupantId": "occupant_ownerA",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 50.55,
      "totalCost": 303.3,
    },
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 450.72,
      "occupantId": "occupant_ownerB",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 75.12,
      "totalCost": 450.72,
    },
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 127.433,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterA",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 1274.33,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 127.433,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterB",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 2548.66,
    },
    {
      "billedArea": 30,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 127.433,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 3822.99,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 127.433,
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 1500,
      "totalCost": 9000,
    },
  ],
  "consumptionRecordsToInsert": [
    {
      "consumption": 100,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_renterWithDevice",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
    {
      "consumption": 50.55,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_ownerA",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
    {
      "consumption": 75.12,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_ownerB",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`calculateBills > handles real world calculation scenario 1`] = `
{
  "billsToInsert": [
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 900,
      "occupantId": "occupant_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 150,
      "totalCost": 900,
    },
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "measuredCost": 1200,
      "occupantId": "occupant_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 200,
      "totalCost": 1200,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 90,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renter1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 1800,
    },
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 90,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renter2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 900,
    },
    {
      "billedArea": 50,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 90,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "occupantId": "occupant_renter3",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 4500,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 90,
      "costPerUnit": 6,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 1550,
      "totalCost": 9300,
    },
  ],
  "consumptionRecordsToInsert": [
    {
      "consumption": 150,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
    {
      "consumption": 200,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "electricity",
      "measuringDeviceId": "measurement_record_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`calculateBills > handles real world calculation scenario 2`] = `
{
  "billsToInsert": [
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 88.22,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "fixedCost": undefined,
      "measuredCost": 264.66,
      "occupantId": "occupant_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 3,
      "totalCost": 264.66,
    },
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 88.22,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "fixedCost": undefined,
      "measuredCost": 573.43,
      "occupantId": "occupant_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 6.5,
      "totalCost": 573.43,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 2.315775,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "occupantId": "occupant_renter1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 46.3155,
    },
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 2.315775,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "occupantId": "occupant_renter2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 23.15775,
    },
    {
      "billedArea": 50,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 2.315775,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "occupantId": "occupant_renter3",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 115.78875,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 2.315775,
      "costPerUnit": 88.22,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 11.6,
      "totalCost": 1023.352,
    },
  ],
  "consumptionRecordsToInsert": [
    {
      "consumption": 3,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "measuringDeviceId": "measurement_record_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
    {
      "consumption": 6.5,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "water",
      "measuringDeviceId": "measurement_record_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`calculateBills > handles real world calculation scenario 3`] = `
{
  "billsToInsert": [
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "fixedCost": undefined,
      "measuredCost": 1967.792105263158,
      "occupantId": "occupant_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 5,
      "totalCost": 1967.792105263158,
    },
    {
      "billingPeriodId": "billing_period_1",
      "costPerUnit": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "fixedCost": 64.02048655569783,
      "measuredCost": 3935.584210526316,
      "occupantId": "occupant_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 10,
      "totalCost": 3999.6046970820134,
    },
    {
      "billedArea": 20,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "occupantId": "occupant_renter1",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 7871.168421052632,
    },
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "occupantId": "occupant_renter2",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 3935.584210526316,
    },
    {
      "billedArea": 50,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "occupantId": "occupant_renter3",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 19677.92105263158,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 393.55842105263156,
      "costPerUnit": 393.55842105263156,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "fixedCost": 5000,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 95,
      "totalCost": 42388.05,
    },
    {
      "billingPeriodId": "billing_period_1",
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "fixedCost": 4935.979513444302,
      "occupantId": "occupant_oliera",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 0,
      "totalCost": 4935.979513444302,
    },
  ],
  "consumptionRecordsToInsert": [
    {
      "consumption": 5,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "measuringDeviceId": "measurement_record_owner1",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
    {
      "consumption": 10,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "measuringDeviceId": "measurement_record_owner2",
      "startDate": 2024-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`calculateBills > handles zero total consumption 1`] = `
{
  "billsToInsert": [
    {
      "billedArea": 10,
      "billingPeriodId": "billing_period_1",
      "costPerSquareMeter": 900,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "occupantId": "occupant_renterA",
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalCost": 9000,
    },
    {
      "billingPeriodId": "billing_period_1",
      "buildingId": "building_1",
      "costPerSquareMeter": 900,
      "costPerUnit": 0,
      "endDate": 2024-01-31T00:00:00.000Z,
      "energyType": "heating",
      "fixedCost": undefined,
      "startDate": 2024-01-01T00:00:00.000Z,
      "totalConsumption": 0,
      "totalCost": 9000,
    },
  ],
  "consumptionRecordsToInsert": [],
}
`;
